packages/worker/src/workflow/reusable-vm.ts

```diff
@@ -104,39 +119,62 @@ export class ReusableVMWorkflowCreator implements WorkflowCreator {
    */
   async createWorkflow(options: WorkflowCreateOptions): Promise<Workflow> {
     const context = this.context;
-    const bag: Record<string, unknown> = {};
-    const { isolateExecutionTimeoutMs, contextKeysToPreserve } = this;
+    const pristine = this.pristine;
+    const bag: Map<string, PropertyDescriptor> = new Map();
+    const { isolateExecutionTimeoutMs } = this;
     const workflowModule: WorkflowModule = new Proxy(
       {},
       {
         get(_: any, fn: string) {
           return (...args: any[]) => {
-            Object.assign(context, bag);
+            for (const [pname, pdesc] of bag.entries()) {
+              Object.defineProperty(context, pname, pdesc);
+            }
+
             // runInContext does not accept args, pass via globals
             context.__TEMPORAL_ARGS__ = args;
             try {
+              // By the time we get out of this call, all microtasks will have been executed
               return vm.runInContext(`__TEMPORAL__.api.${fn}(...__TEMPORAL_ARGS__)`, context, {
                 timeout: isolateExecutionTimeoutMs,
                 displayErrors: true,
               });
             } finally {
-              const keysToDelete = [];
-              // TODO: non-enumerable global properties?
-              // https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects
-              for (const k in context) {
-                if (!contextKeysToPreserve.has(k)) {
-                  bag[k] = context[k];
-                  context[k] = undefined;
-                  keysToDelete.push(k);
+              // No need to keep that one
+              delete context.__TEMPORAL_ARGS__;
+
+              //
+              bag.clear();
+              const keysToCleanup: [string, PropertyDescriptor | undefined][] = [];
+
+              for (const pname of Object.getOwnPropertyNames(context)) {
+                const pdesc = Object.getOwnPropertyDescriptor(context, pname) as PropertyDescriptor;
+                const pdescPristine = pristine?.get(pname);
+                if (pdesc.value !== pdescPristine?.value) {
+                  bag.set(pname, pdesc);
+                  keysToCleanup.push([pname, pdescPristine]);
                 }
               }
-              for (const k in keysToDelete) {
-                delete context[k];
-              }
+
+              // Looks like Node/V8 is not properly syncing deletion of keys on the outter context
+              // object to the inner globalThis object. Hence, we delete them from inside the context.
+              context.__TEMPORAL_ARGS__ = keysToCleanup;
+              vm.runInContext(
```

Comment:
This will add a bit of overhead for every time we enter the context, I would try to find an alternative way to do this.